### PR TITLE
Rewrite/refactor delagent and fix #273

### LINKED ISSUE #273: any user who is not the owner can delete any folder via /delagent -F
---

Author Name: **larry shi** 
Original Redmine Issue: 6741, http://www.fossology.org/issues/6741
Original Date: 2014/03/18
Original Assignee: Dong Ma

---

None

### PR DESCRIPTION
**Pull request by maximilianhuber**
This PR 
* adds permission handling for directories
* refactors delagent
* repairs permission checking

this will fix #273
see #602 for the discussion on the previous PR concerned with this issue

--
This PR should not be merged yet.

### CODE REVIEW

**Comment by maximilianhuber** [comment]
@shaheemazmalmmd: even with the debugger i was not able to find the difference in 
```c
[..]
PGresult * PQexecCheck(const char *desc, char *SQL, char *file, const int line)
{
  PGresult *result;
  [...]
  result = PQexec(db_conn, SQL);
  if (fo_checkPQcommand(db_conn, result, SQL, file, line))
  {
    exit(-1);
  }
  return result;
}
[..]
#if 0
  result = PQexecCheck(NULL, SQL, __FILE__, __LINE__);
#else
  result = PQexec(db_conn, SQL);
  if (fo_checkPQresult(db_conn, result, SQL, __FILE__, __LINE__))
  {
    exit(-1);
  }#if 0
  result = PQexecCheck(NULL, SQL, __FILE__, __LINE__);
#else
  result = PQexec(db_conn, SQL);
  if (fo_checkPQresult(db_conn, result, SQL, __FILE__, __LINE__))
  {
    exit(-1);
  }
#endif
#endif
[...]
```
I would suggest to remove the `#if 0` case.

Btw.: I would rebase this PR if this is OK for you?

**Comment by maximilianhuber** [comment]
@shaheemazmalmmd: I have already rebased it but I can revert it if you have problems with that.

**Comment by shaheemazmalmmd** [comment]
no,  good that you rebased it. i missed it yesterday.

**Comment by maximilianhuber** [comment]
I think that we have to have a look at the tests for delagent. I think at least the functional agent tests are outdated and also the unit tests should be extended.

There is currently also a logic problem in the code where uploads contained in a Folder get deleted, even without write permissions. I will try to fix this.

**Comment by maximilianhuber** [comment]
### Problem:
`./delagent --user rUser --password --rPassword -f`, where `rUser` is a user with read permissions, lists **all uploads and ignores permissions**

On the other side does `./delagent [...] -u` only list uploads where one has read permissions.

### Solution:
use the logic/permission checks from `-u`

**Comment by maximilianhuber** [comment]
@shaheemazmalmmd: I have manually tested that PR and it looks OK. Do you have any ideas what could be missing? Otherwise I would ask @steffen-weber or @wuan to review this PR. After that I might squash the commits and remove that silly `#if 0` thing.

BUT: I am still not able to run the functional tests. The used scripts `src/testing/lib/libTestDB.php` and `src/testing/db/gtdbcreate.sh` seam to be outdated. I think they should be replaced by use `Fossology\Lib\Test\TestPgDb` together with `Fossology\Lib\Test\TestInstaller`. But in this environment one has no deployed `www` and thus cant use the upload functionality to set the DB up.

**Comment by maximilianhuber** [comment]
Hello, 

we need someone to review this PR. Maybe @wuan or @steffen-weber? Or maybe even @fogninid  since this is essentially c-code?

When someone volunteers, I will support him actively and give an overview on the changes.

**Comment by maximilianhuber** [comment]
Oh sorry, instead of @fogninid I wanted to write @joshovi.

File: src/delagent/agent/delagent.c [hunk 1]
```diff
@@ -31,11 +32,59 @@ char BuildVersion[]="delagent build version: NULL.\n";
 #endif
 
 
+/***********************************************
+ Usage():
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: I would prefer all function names and variable names to be start with lower case. It should be uniform at least.

File: src/delagent/agent/delagent.c [hunk 2]
```diff
@@ -77,8 +126,6 @@ int main (int argc, char *argv[])
   int Scheduler=0; /* should it run from the scheduler? */
   int GotArg=0;
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: why not gotArg

File: src/delagent/agent/delagent.c [hunk 3]
```diff
@@ -77,8 +126,6 @@ int main (int argc, char *argv[])
   int Scheduler=0; /* should it run from the scheduler? */
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: scheduler

File: src/delagent/agent/delagent.c [hunk 4]
```diff
@@ -77,8 +126,6 @@ int main (int argc, char *argv[])
...
   int GotArg=0;
   char *agent_desc = "Deletes upload.  Other list/delete options available from the command line.";
-  char *Parm = NULL;
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: *parm

File: src/delagent/agent/util.c [hunk 5]
```diff
@@ -524,8 +531,11 @@ void DeleteUpload (long UploadId)
  * \param long Row grandparent (used to unlink if multiple grandparents)
  * \param int DelFlag 0=no del, 1=del if unique parent, 2=del unconditional
  *
+ * \return 1: success;
+ *         0: fail
```
Comment (joshovi) [inline src/delagent/agent/util.c]: That is the opposite of the usual convention for return codes

File: src/delagent/agent/util.c [hunk 6]
```diff
@@ -242,280 +419,110 @@ void DeleteUpload (long UploadId)
...
+  return 1; /* success */
+} /* DeleteUpload() */
 
-  result = PQexec(db_conn, "SET statement_timeout = 120000;");
-  if (fo_checkPQcommand(db_conn, result, "SET statement_timeout = 120000;", __FILE__, __LINE__)) exit(-1);
-  PQclear(result);
+/**
+ * \brief remove link between parent and (child,mode) if there are other parents
+ *
+ * \return 1: successfully deleted link (other link existed);
```
Comment (joshovi) [inline src/delagent/agent/util.c]: That is the opposite of the usual convention for return codes

File: src/delagent/agent/util.c [hunk 7]
```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
...
+ * \brief check if the upload can be deleted, that is the user have
+ * the permission to delete this upload
+ *
+ * \param long upload_id - upload id
+ * \param char *user_name - user name
+ *
+ * \return 1: yes, you have the needed permissions;
+ *         0: no;
+ *        -1: failure;
+ *        -2: does not exist
```
Comment (joshovi) [inline src/delagent/agent/util.c]: That is an unusual convention for return codes

File: src/delagent/agent/delagent.h [hunk 8]
```diff
@@ -42,18 +43,29 @@ extern PGconn* db_conn;
...
 int authentication(char *user, char * password, int *user_id, int *user_perm);
-int check_permission_del(long upload_id, int user_id, int user_perm);
+
+int check_permission_upload(int wantedPermissions, long upload_id, int user_id, int user_perm);
+int check_read_permission_upload(long upload_id, int user_id, int user_perm);
+int check_write_permission_upload(long upload_id, int user_id, int user_perm);
+int check_permission_folder(long folder_id, int user_id, int user_perm);
+int check_permission_license(long license_id, int user_perm);
+
+/* functions that list things */
```
Comment (joshovi) [inline src/delagent/agent/delagent.h]: does the captital letter at the start signify anything?

File: src/delagent/agent/util.c [hunk 9]
```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
...
+/**
+ * \brief check if the upload can be deleted, that is the user have
+ * the permission to delete this upload
+ *
+ * \param long upload_id - upload id
+ * \param char *user_name - user name
+ *
+ * \return 1: yes, can be deleted;
+ *         0: can not be deleted;
+ *        -1: failure;
```
Comment (joshovi) [inline src/delagent/agent/util.c]: return code convention

File: src/delagent/agent/util.c [hunk 10]
```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
...
+/**
+ * \brief check if the upload can be deleted, that is the user have
+ * the permissoin to delete this upload
+ *
+ * \param long upload_id - upload id
+ * \param char *user_name - user name
+ *
+ * \return 1: yes, can be deleted;
+ *         0: can not be deleted;
+ */
```
Comment (joshovi) [inline src/delagent/agent/util.c]: see above return code convention

File: src/delagent/agent/util.c [hunk 11]
```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
...
+int check_write_permission_license(long license_id, int user_perm)
+{
+  if (user_perm != PERM_ADMIN)
+  {
+    verbosePrintf("only admin is allowed to delete licenses\n");
```
Comment (joshovi) [inline src/delagent/agent/util.c]: this is not a self explanatory name. That would imply that it adds more information than a normal printf

File: src/delagent/agent/delagent.h (hunk 8 above)
Comment (maximilianhuber) [inline src/delagent/agent/delagent.h]: No, that is just the style the code was written...

File: src/delagent/agent/util.c (hunk 7 above)
Comment (maximilianhuber) [inline src/delagent/agent/util.c]: Do you have a better Idea?

An alternative would be to use:
* `0` for *yes*
* `1` for *no* and *failure*

Then there is the problem that *does not exist* is not always a failure, but I'm currently not sure whether this differentiation is still needed.

Comment (joshovi) [inline src/delagent/agent/util.c]: well 0 for no problem is what the main function should return, and why should functions do it differently?
Anything != 0 would be some kind of error code.

File: src/delagent/agent/util.c [hunk 12]
```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
...
+  char SQL[MAXSQL] = {0};
+  PGresult *result;
+  char user_seed[myBUFSIZ] = {0};
+  char pass_hash_valid[41] = {0};
+  unsigned char pass_hash_actual_raw[21] = {0};
+  char pass_hash_actual[41] = {0};
+
+  /** get user_seed, user_pass on one specified user */
+  snprintf(SQL,MAXSQL,"SELECT user_seed, user_pass, user_perm, user_pk from users where user_name='%s';", user);
+  result = PQexec(db_conn, SQL);
```
Comment (joshovi) [inline src/delagent/agent/util.c]: Remind me why little Bobby tables cannot cause a problem here

File: src/delagent/agent/delagent.c [hunk 13]
```diff
@@ -131,49 +208,56 @@ int main (int argc, char *argv[])
...
-  
-  if (ListProj) ListUploads(user_id, user_perm);
-  if (ListFolder) ListFolders(user_id);
-
-  alarm(60);  /* from this point on, handle the alarm */
-  if (DelUpload) 
-  {
-    if (1 != check_permission_del(DelUpload, user_id, user_perm))
+    if (1 != authentication(user_name, password, &user_id, &user_perm))
     {
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: Remind me why little Bobby tables cannot cause a problem here

**Comment by maximilianhuber** [comment]
I will refactor the returncodes tomorrow

File: src/delagent/agent/delagent.c [hunk 14]
```diff
@@ -96,84 +140,124 @@ int main (int argc, char *argv[])
...
+    fo_GetAgentKey(db_conn, basename(argv[0]), 0, agent_rev, agent_desc);
+
+    if (listProj)
     {
-      Parm = fo_scheduler_current();
-      
-      if (ReadParameter(Parm) < 0)
-        exit(-1);
+      returnedCode = listUploads(user_id, user_perm);
+    }
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: if returnedCode !=  0 you might want to leave the program here before you set it again if listFolder is true

Comment (maximilianhuber) [inline src/delagent/agent/delagent.c]: You are right. Its fixed now

File: src/delagent/agent/delagent.c [hunk 15]
```diff
@@ -96,84 +140,128 @@ int main (int argc, char *argv[])
...
+
+    alarm(60);  /* from this point on, handle the alarm */
+    if (delUpload)
+    {
+      returnedCode = deleteUpload(delUpload, user_id, user_perm);
+
+      writeMessageAfterDelete("upload", delUpload, user_name, returnedCode);
+    }
+    if (delFolder)
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: same here, or are delUpload and delFolder mutually exclusive?

File: src/delagent/agent/delagent.c [hunk 16]
```diff
@@ -96,84 +140,128 @@ int main (int argc, char *argv[])
...
+    {
+      returnedCode = deleteFolder(delFolder, user_id, user_perm);
+
+      writeMessageAfterDelete("folder", delFolder, user_name, returnedCode);
+    }
+    if (delLicense)
```
Comment (joshovi) [inline src/delagent/agent/delagent.c]: ^^

File: src/delagent/agent/delagent.c (hunk 15 above)
Comment (maximilianhuber) [inline src/delagent/agent/delagent.c]: The function `writeMessageAfterDelete` keeps track of that and exits if the returned code is not `0`

File: src/delagent/ui/admin-folder-delete.php [hunk 17]
```diff
@@ -83,19 +83,24 @@ public function Output() {
     /* If this is a POST, then process the request. */
```
Comment (joshovi) [inline src/delagent/ui/admin-folder-delete.php]: I forgot much already, but I remember that you can port the plugins to the new rendering mechanism with a template. Depending on how much work this is, it's always nice to separate work/logic from the template

File: src/delagent/ui/admin-upload-delete.php [hunk 18]
```diff
@@ -87,19 +113,7 @@ public function Output()
...
-        $V.= displayMessage($msg);
-      }
-      else {
-        $text=_("Deletion Scheduling failed: ");
-        $V.= DisplayMessage($text.$rc);
-      }
+      $V.= $this->TryToDelete($uploadpk);
     }
     /* Create the AJAX (Active HTTP) javascript for doing the reply
      and showing the response. */
```
Comment (joshovi) [inline src/delagent/ui/admin-upload-delete.php]: again, if you find the time use the templates

File: src/delagent/ui/admin-folder-delete.php (hunk 17 above)
Comment (maximilianhuber) [inline src/delagent/ui/admin-folder-delete.php]: Yes, I support this. But I would move that to another PR.

File: src/lib/c/libfossagent.c [hunk 19]
```diff
@@ -271,6 +271,58 @@ FUNCTION int fo_CreateARSTable(PGconn* pgConn, const char* tableName)
...
+  result = PQexec(pgConn, SQL);
+  fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__);
+  if (!fo_checkPQresult(pgConn, result, SQL, __FILE__, __LINE__)
+    && PQntuples(result) > 0)
+  {
+    permPublic = atoi(PQgetvalue(result, 0, 0));
+  }
+  PQclear(result);
+
+  permExternal = ( permGroup > permPublic ) ? permGroup : permPublic;
```
Comment (joshovi) [inline src/lib/c/libfossagent.c]: this could be a function like 

int max(int permGroup, int permPublic)

File: src/lib/c/libfossagent.c [hunk 20]
```diff
@@ -271,6 +271,58 @@ FUNCTION int fo_CreateARSTable(PGconn* pgConn, const char* tableName)
...
+  return ( user_perm < permExternal ) ? user_perm: permExternal;
```
Comment (joshovi) [inline src/lib/c/libfossagent.c]: and this 

int min(int user_perm , int permExternal)

File: src/lib/c/libfossagent.h [hunk 21]
```diff
@@ -31,6 +31,7 @@ int fo_GetAgentKey(PGconn* pgConn, const char* agent_name, long unused, const ch
 int fo_WriteARS(PGconn* pgConn, int ars_pk, int upload_pk, int agent_pk,
   const char* tableName, const char* ars_status, int ars_success);
 int fo_CreateARSTable(PGconn* pgConn, const char* table_name);
+int GetUploadPermP(PGconn* pgConn, long UploadPk, int user_pk, int user_perm);
```
Comment (joshovi) [inline src/lib/c/libfossagent.h]: can you add a more descriptive name?

**Comment by maximilianhuber** [comment]
@joshovi: your last mentioned issues are solved with the commit 4958b92. When merging this should be squashed into 16764981284fe6b64e4c91ab1d0b1e1ff4b5d278

File: src/lib/c/libfossagent.h (hunk 21 above)
Comment (maximilianhuber) [inline src/lib/c/libfossagent.h]: I have already thought about that. I had no good idea yet.
